src/python/pants/backend/python/providers/pyenv/rules.py

```diff
@@ -0,0 +1,183 @@
+        help="Flags to use when configuring CPython.",
+        advanced=True,
+    )
+
+    @classproperty
+    def default_known_versions(cls):
+        return [
+            "|".join(
+                (
+                    cls.default_version,
+                    plat,
+                    "9105de5e5cf8dc0eca2a520ed04493d183128d46a2cfb402d4cc271af1bf144b",
+                    "749323",
+                )
+            )
+            for plat in ["macos_arm64", "macos_x86_64", "linux_x86_64", "linux_arm64"]
+        ]
+
+    def generate_exe(self, plat: Platform) -> str:
+        """Returns the path to the tool executable.
+
+        If the downloaded artifact is the executable itself, you can leave this unimplemented.
+
+        If the downloaded artifact is an archive, this should be overridden to provide a
+        relative path in the downloaded archive, e.g. `./bin/protoc`.
+        """
+        return f"./pyenv-{self.version}/bin/pyenv"
+
+
+class PythonToolchainRequest(PythonProvider):
+    pass
+
+
+PYENV_NAMED_CACHE = ".pyenv"
+PYENV_APPEND_ONLY_CACHES = FrozenDict({"pyenv": PYENV_NAMED_CACHE})
+
+
+@rule
+async def get_python(
+    request: PythonToolchainRequest,
+    python_setup: PythonSetup,
+    platform: Platform,
+    pyenv_subsystem: PyenvSubsystem,
+) -> PythonExecutable:
+    env_vars, pyenv = await MultiGet(
+        Get(EnvironmentVars, EnvironmentVarsRequest(["PATH", "LDFLAGS"])),
+        Get(DownloadedExternalTool, ExternalToolRequest, pyenv_subsystem.get_request(platform)),
+    )
+
+    python_to_use = request.interpreter_constraints.minimum_python_version(
+        python_setup.interpreter_versions_universe
+    )
+    if python_to_use is None:
+        raise ValueError(
+            f"Couldn't determine a compatible Interpreter Constraint from {python_setup.interpreter_versions_universe}"
+        )
+
+    which_python_result = await Get(
+        ProcessResult,
+        Process(
+            [pyenv.exe, "latest", "--known", python_to_use],
+            input_digest=pyenv.digest,
+            description=f"Choose specific version for Python {python_to_use}",
+            env={"PATH": env_vars.get("PATH", "")},
+            # Caching the result is OK, since if the user really needs a different patch,
+            # they should list a more precise IC.
+        ),
+    )
+    specific_python = which_python_result.stdout.decode("ascii").strip()
+
+    shim_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "install_python_shim.sh",
+                    dedent(
+                        f"""\
+                        #!/usr/bin/env bash
+                        set -e
+                        DEST=$(realpath {PYENV_NAMED_CACHE}/{specific_python})
+                        if [ ! -d "$DEST" ]; then
+                            export PYENV_ROOT=$(realpath $(mktemp -d -u -p {PYENV_NAMED_CACHE} {specific_python}.XXXXXX))
```

Comment:
OK so we create an ad-hoc PYENV_ROOT for each install (which is why specific_python is mixed into its name)? That is not idiomatic for pyenv, which normally keeps all its versions in the same root, but it makes sense here I think.

Reply:
Yup. And that is because:

> Python installation is not atomic.

So if we did just set `PYENV_ROOT` to the named cache dir, the resulting directory/binary would exist before its "ready".
Alternatively we could `touch PYENV_ROOT/versions/{version}/DONE` but that means `pyenv` installations in parallel have to be OK. They might be but IDK if they are :unamused: 

I'm very eager to make this cleaner, so feel free to brainstorm as much as you want with me :smile:

Reply:
Eh, I think it's fine... :)